From 559dbfd5c3a571416c9ad60de81401a6a506dea7 Mon Sep 17 00:00:00 2001
From: Udayan Banerji <udayan.banerji@intel.com>
Date: Wed, 20 Mar 2013 11:19:07 -0700
Subject: Dalvik: Fix propagation of errors from back-end

BZ: 92503

1. Fixes returning of error from collectInfoOfBasicBlock.
2. lowerByteCodeJit can fail for reasons other than unsupported bytecode.
   Fix that assumption.

Category: device-enablement
Domain: AOSP-Dalvik-Compiler-CG
Origin: internal
Upstream-Candidate: no, needs rework

Change-Id: Ib1084738ac83d550897e027d3bc36e9bffad3d74
Orig-MCG-Change-Id: I4dd1980055f569a445d0b725bc2e8cbc6e32e949
Signed-off-by: Udayan Banerji <udayan.banerji@intel.com>
Signed-off-by: Qiming Shi <qiming.shi@intel.com>
Signed-off-by: Serguei Katkov <serguei.i.katkov@intel.com>
---
 vm/compiler/codegen/x86/AnalysisO1.cpp       |    6 +++++-
 vm/compiler/codegen/x86/CodegenInterface.cpp |    6 ++++--
 2 files changed, 9 insertions(+), 3 deletions(-)

diff --git a/vm/compiler/codegen/x86/AnalysisO1.cpp b/vm/compiler/codegen/x86/AnalysisO1.cpp
index 8f3c116..df8cd0b 100644
--- a/vm/compiler/codegen/x86/AnalysisO1.cpp
+++ b/vm/compiler/codegen/x86/AnalysisO1.cpp
@@ -888,7 +888,11 @@ int preprocessingBB(CompilationUnit *cUnit, BasicBlock* bb) {
     if (bb->blockType != kDalvikByteCode && bb->firstMIRInsn == NULL)
         return 0;
 
-    collectInfoOfBasicBlock(currentMethod, currentBB);
+    int retCode = collectInfoOfBasicBlock(currentMethod, currentBB);
+    if (retCode < 0) {
+        return retCode;
+    }
+
 #ifdef DEBUG_COMPILE_TABLE
     dumpVirtualInfoOfBasicBlock(currentBB);
 #endif
diff --git a/vm/compiler/codegen/x86/CodegenInterface.cpp b/vm/compiler/codegen/x86/CodegenInterface.cpp
index 5bdb687..fdff23b 100644
--- a/vm/compiler/codegen/x86/CodegenInterface.cpp
+++ b/vm/compiler/codegen/x86/CodegenInterface.cpp
@@ -1518,9 +1518,11 @@ static bool generateCode (CompilationUnit *cUnit, BasicBlock *bb, BasicBlock **n
 
             //If not handle, error flag setting and return false
             if (notHandled) {
-                ALOGI("JIT_INFO: Opcode 0x%x (%s) / Fmt %d at offset %#06x not handled\n",
+                ALOGI("JIT_INFO: Opcode 0x%x (%s) / Fmt %d at offset %#06x not lowered\n",
                         dalvikOpCode, dexGetOpcodeName(dalvikOpCode), dalvikFormat, mir->offset);
-                SET_JIT_ERROR(kJitErrorUnsupportedBytecode);
+                //Lowering could have failed due to various reasons.
+                //Set a generic error flag
+                SET_JIT_ERROR(kJitErrorCodegen);
                 cUnit->baseAddr = NULL;
                 endOfTrace(true); /* need to free structures */
                 return false;
-- 
1.7.4.1

